design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+The required boot mode for a host depends on several factors,
+including support in the hardware, whether IPv6 is being used, and
+whether virtual media is being used. We do not need to support all
+combinations of those features with all boot modes. Asking the user to
+give us the boot mode explicitly introduces an opportunity for them to
+specify an incorrect value. We have enough information about the host
+already to determine the correct boot mode to use. Therefore we should
+not need to extend the BareMetalHost API in order to support multiple
+boot modes.
+
+## Motivation
+
+When booting using virtual media, Ironic builds a custom ISO for each
+host. Today, it only includes the correct contents for one boot mode
+at a time. Some attempt is made to determine the right value by
+examining the hardware, but that is not always possible. When
+necessary, there is a way to configure a default.
+
+Several proposals have been floated to introduce a way for the user to
+explicitly set a boot mode. This proposal asserts that we can
+successfully manage the x86 hardware supported by metal3 today without
+it. Avoiding a new API field avoids several opportunities for
+misconfiguration.
+
+### Goals
+
+* Describe a method for providing a deterministic boot mode value to
+  Ironic without relying on the user to set the value directly.
+* Ensure that metal3's behavior is always deterministic.
+
+### Non-Goals
+
+* Address boot mode selection for non-x86 hardware.
+
+## Proposal
+
+For each host, Ironic maintains a set of properties telling it how to
+manage the hardware. One of these is the `boot_mode`. Valid values
+are `bios` and `uefi`. When the `boot_mode` property is
+explicitly set, Ironic will try to ensure that the hardware is
+configured accordingly and it may also choose different paths in the
+provisioning process. In particular, there is a branch in the ISO
+building logic to include different content based on the boot mode.
+
+If the `boot_mode` is not specified, Ironic tries to determine the
+right value by looking at the hardware settings. Some BMC types do not
+support this capability, especially when using the IPMI protocol. When
+Ironic cannot determine the right value for itself, and has not been
+told explicitly what value to use, it falls back to a global default
+specified by a configuration option. Metal3 is not typically deployed
```

Comment:
What is this global default fallback? Is it simple as ```BIOS``` or ```UEFI``` or something else?

Reply:
basically "bios" or "uefi" if memory serves. The default ironic ships is bios, but we've got notes and reminders in our code base to change that to UEFI. The only thing really holding that back (If there were infinite time/humans working on the code) is IPMI and risk of breaking users.

Reply:
I see. And having this default as ```bios```, could it complicate things for us? In this specific context of the operator.

Reply:
The default is configurable, so we could change it in the copy of Ironic that we deploy if what is being used now turns into a problem.

Reply:
Sounds good.